util/include/throughput.hpp

```diff
@@ -17,56 +17,78 @@
  * If the last call to the tracker is start(): call totalDuration() to get the sum of intervals, including current
  * still running interval.
- * If the last call to a tracker is pause(): you may get the sum of intervals from pause() returned
+ * If the last call to a tracker is stop(): caller may get the sum of intervals from stop() returned
  * value or call totalDuration() explicitly.
- * To reset (re-use) the tracker, call reset() and then start().
+ * To reset (re-use) the tracker, call start/stop with do_reset=true.
  */
 template <typename T>
 class DurationTracker {
  public:
-  void start() {
-    ConcordAssert(!running_);
+  DurationTracker(std::string&& name = "", bool do_start = false)
+      : total_duration_{}, start_time_{}, running_{do_start}, name_(name) {
+    static size_t obj_counter{};
+    if (do_start) {
+      start_time_ = std::chrono::steady_clock::now();
+    }
+    if (name.empty()) {
+      name_ = std::to_string(++obj_counter);
+    }
+    LOG_INFO(GL, KVLOG(name_, do_start));
+  }
+
+  void start(bool do_reset = false) {
+    ConcordAssertOR(running_, !do_reset);
     start_time_ = std::chrono::steady_clock::now();
+    if (do_reset) {
+      total_duration_ = 0;
+    }
     running_ = true;
+    LOG_INFO(GL, KVLOG(name_, running_, total_duration_, do_reset));
   }
-  uint64_t pause() {
-    if (running_)
+
+  uint64_t stop(bool do_reset = false) {
+    if (running_ && !do_reset) {
       total_duration_ += std::chrono::duration_cast<T>(std::chrono::steady_clock::now() - start_time_).count();
+    }
+    uint64_t ret = total_duration_;
+    if (do_reset) {
+      total_duration_ = 0;
+      start_time_ = TimePoint();
+    }
     running_ = false;
-    return total_duration_;
-  }
-  void reset() {
-    total_duration_ = 0;
-    running_ = false;
-  }
-  void restart() {
-    start_time_ = std::chrono::steady_clock::now();
-    running_ = true;
+    LOG_INFO(GL, KVLOG(name_, running_, total_duration_, ret, do_reset));
+    return ret;
   }
-  uint64_t totalDuration(bool doReset = false) {
-    uint64_t ret = total_duration_;
+
+  // when do_reset is true, tracker keeps its state (running/not running), returns total_duration_ up to current point
+  // in time and resets total_duration_ to 0
+  uint64_t totalDuration(bool do_reset = false) {
+    uint64_t ret;
     if (running_) {
-      total_duration_ = pause();
-      ret = total_duration_;
-      if (doReset) {
-        reset();
-      } else
-        start();
+      total_duration_ = std::chrono::duration_cast<T>(std::chrono::steady_clock::now() - start_time_).count();
+    }
+    ret = total_duration_;
+    if (do_reset) {
+      total_duration_ = 0;
     }
+    LOG_INFO(GL, KVLOG(name_, running_, total_duration_, ret, do_reset));
     return ret;
   };
 
- private:
+ protected:
   uint64_t total_duration_ = 0;
-  std::chrono::time_point<std::chrono::steady_clock> start_time_;
+  using TimePoint = std::chrono::time_point<std::chrono::steady_clock>;
```

Comment:
I think all time related declarations should go at corresponding header file.

Reply:
Actually, this is a very bad thing to have using-drectives in an hpp.
So much was written about it.
In practice we should use it only in class or function scope.
Relaying on a using-directive of an included header is not a good idea as well.
At most, we should declare it on the top of our cpp - even in that case it is not a good thing to do (user directive at cpp file scope).
To summarize from https://www.learncpp.com/cpp-tutorial/using-declarations-and-using-directives:
"Prefer explicit namespaces over using statements. Avoid using directives whenever possible. Using declarations are okay to use inside blocks."